lib/semantics/resolve-names.cc

```diff
@@ -4086,9 +4085,14 @@ bool ConstructVisitor::Pre(const parser::LocalitySpec::LocalInit &x) {
 
 bool ConstructVisitor::Pre(const parser::LocalitySpec::Shared &x) {
   for (const auto &name : x.v) {
-    Symbol *prev{EnsureEnclosingEntity(name)};
-    if (PassesSharedLocalityChecks(name, *prev)) {
-      auto &symbol{MakeSymbol(name, HostAssocDetails{*prev})};
+    if (!FindSymbol(name)) {
+      Say(name,
+          "Warning, variable '%s' with SHARED locality allocated"
+          " automatically and may be uninitialized"_en_US);
```

Comment:
We don't start other warnings with "Warning," so we shouldn't here either. When issue #476 is implemented we will have a way to mark warnings and then they should come out with "warning:" before the message.

I would say "implicitly declared" rather than "allocated automatically". It's not allocated in the Fortran sense.

Reply:
And it definitely will be uninitialized within the loop, and probably uninitialized afterwards -- and if it's not used after the loop, it shouldn't have been shared anyway.  Maybe this whole feature is a bad idea.

Reply:
Thanks, @tskeith.  I'll change the message.

Reply:
@vdonaldson, it was based on information from you in an email message that made me think it was a good idea to allow for implicitly declared variables to be named in a SHARED locality spec.  But I agree with @klausler that it's hard to imagine why anyone would want to do this.  Do you have any insight into this?

Reply:
Do concurrent locality lists are new in f18, and not necessarily well thought out or well designed.   Constraint C1124 does not distinguish between local, local_init, and shared sub-cases.  If implicit declaration of local and local_init variables is allowed, then implicit declaration of shared variables is as well, at least with regard to C1124.  As Peter notes, this program idiom is probably standard conforming, but useless:

```
    do concurrent (i=1:10) shared(s)
       ! code that doesn't reference s
    end do
```

The following code also doesn't violate C1124.  It is non-conforming from 11.1.7.5p4 because it has assignments to shared variable 's' in multiple iterations.  The compiler is not required to diagnose this violation. It should not mark the 'shared(s)' locality spec as an error, because that is not the error.  It would be ok to flag the assignment as an error, although that may be beyond current compiler capabilities.

```
    do concurrent (i=1:10) shared(s)
      s = 1.0
      ! do something with s
    end do

```

Short of changes to the standard or clarifying references elsewhere (unlikely), it looks like this
implicit declaration should be treated as a usual implicit declaration.  Two secondary responses seem reasonable:

 - Do nothing (else); this is not a constraint violation.
 - Generate a warning, possibly as neutral as "Implicit declaration of '%s' in a SHARED locality spec", although it might be possible to be more informative about why that is questionable.

Note that subscripting an implicitly declared locality spec name will probably generate some error on use.

Reply:
Thanks, Val.

I assume that your reference to the standard was to section 11.1.7.5p3, which discusses SHARED locality.  The more I learn and read about this, the crazier it seems.  This paragraph says (among other things), referring to the variable in the SHARED locality-spec: "If it is defined or becomes undefined during any iteration, it shall not be referenced, defined, or become undefined during any other iteration."  So the only valid use that I can think of would be for one of the iterations to define the variable and then for code outside the DO construct to use the defined value.

Reply:
Yes, I meant 11.1.7.5p3.  And yes, Peter K is right about the usefulness of allowing an implicit declaration in a shared locality spec, and about the design of the feature.  Do concurrent is more or less a replacement for a broken forall construct/statement, which is now deprecated.  Maybe the committee will need a third attempt to get it more or less right.

Reply:
Thanks, Val.

It seems to me that allowing implicit declaration of a name in SHARED locality-spec means that we're allowing people to do something that we don't recommend and can't effectively test.  Thus, even though it seems like the standard allows this, I recommend that we leave things as they are.  That is, that we emit an error message when an undeclared name appears in a SHARED locality-spec.

I'll pursue this problem with the standard outside of this forum.

If no one objects to this plan, I'll withdraw this pull request.

Reply:
I actually don't think that FORALL is all that badly broken, it was just never well implemented.  DO CONCURRENT's default automatic localization rules are what really deserve the term; and the locality specs in 2018 were meant to fix them, and didn't.  They may have been the "third attempt", actually!  Sorry, this is getting totally off-topic for this code review.  Anyway, can you please emit a warning when implicitly declaring a SHARED name?

Reply:
Whether we admit implicit declarations of SHARED scalars or not, we also have the question of whether we should emit a warning about implicit declarations of scalars in DO CONCURRENT that have default locality.

I think that you (Pete) should retain this PR -- one could write a valid Fortran program that would depend on implicit declaration, however artificial.  But also look into coughing up a warning in name resolution when there's an implicit declaration in a DO CONCURRENT without a locality spec.

Reply:
I'm ok with a warning.  I vote against an error.

Reply:
This pull request currently causes the following warning to be issued for undeclared names appearing in a SHARED locality-spec:
```Variable 'a' with SHARED locality implicitly declared and may be uninitialized```

I've added @gklimowicz as a reviewer to see if he has any insight into this.

Reply:
> 
> 
> Whether we admit implicit declarations of SHARED scalars or not, we also have the question of whether we should emit a warning about implicit declarations of scalars in DO CONCURRENT that have default locality.
> 
Thanks, Peter.  I've added the requirement to warn for implicit declaration of variables with unspecified locality to my list on Google Drive.